From 3c8478f6c6da0c02ec19b6d4ad15e9f69bf4bada Mon Sep 17 00:00:00 2001
From: philippe44 <philippe44@users.noreply.github.com>
Date: Fri, 21 Oct 2022 01:37:39 -0700
Subject: [PATCH 1/3] fix can't rename file

rename() will fail at least on ranlib (libtool.c) when used on non-local filesystems because the memory map was not released
---
 cctools/libstuff/ofile.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/cctools/libstuff/ofile.c b/cctools/libstuff/ofile.c
index 499fe798..7dd2bc62 100644
--- a/cctools/libstuff/ofile.c
+++ b/cctools/libstuff/ofile.c
@@ -1497,6 +1497,7 @@ struct ofile *ofile)
 			      "%s", ofile->file_name);
 	    }
 	}
+	munmap(ofile->file_addr, ofile->file_size);
 	if(ofile->file_name != NULL)
 	    free(ofile->file_name);
 	if(ofile->arch_flag.name != NULL)

From 3c08e44023332f987dbb99f42345520ef70d6096 Mon Sep 17 00:00:00 2001
From: philippe44 <philippe44@users.noreply.github.com>
Date: Fri, 21 Oct 2022 11:05:39 -0700
Subject: [PATCH 2/3] was wrong PR

---
 cctools/libstuff/ofile.c | 2 +-
 cctools/misc/libtool.c   | 5 +++++
 2 files changed, 6 insertions(+), 1 deletion(-)

diff --git a/cctools/libstuff/ofile.c b/cctools/libstuff/ofile.c
index 7dd2bc62..ae6675af 100644
--- a/cctools/libstuff/ofile.c
+++ b/cctools/libstuff/ofile.c
@@ -1496,8 +1496,8 @@ struct ofile *ofile)
 		my_mach_error(r, "Can't vm_deallocate mapped memory for file: "
 			      "%s", ofile->file_name);
 	    }
+		munmap(ofile->file_addr, ofile->file_size);		
 	}
-	munmap(ofile->file_addr, ofile->file_size);
 	if(ofile->file_name != NULL)
 	    free(ofile->file_name);
 	if(ofile->arch_flag.name != NULL)
diff --git a/cctools/misc/libtool.c b/cctools/misc/libtool.c
index 22c824d3..bff49bb3 100644
--- a/cctools/misc/libtool.c
+++ b/cctools/misc/libtool.c
@@ -3172,6 +3172,11 @@ struct ofile *ofile)
 	 * Move the temporary file into its final location
 	 */
 	if (tempfile) {
+		/*
+		 * Must release mmap otherwise rename will fail on non-local fs. 
+		 * NB: the ofile_unmap can be made by caller as well
+		 */ 
+		ofile_unmap(ofile);
 	    if (rename(tempfile, output)) {
 		system_fatal("can't move the output file to its final location: %s",
 			     output);

From 0263ba757426b830b437a530fe137be7417377ea Mon Sep 17 00:00:00 2001
From: philippe44 <philippe44@users.noreply.github.com>
Date: Fri, 21 Oct 2022 22:07:16 -0700
Subject: [PATCH 3/3] check for ofile

---
 cctools/misc/libtool.c | 18 ++++++++++--------
 1 file changed, 10 insertions(+), 8 deletions(-)

diff --git a/cctools/misc/libtool.c b/cctools/misc/libtool.c
index bff49bb3..901cc32d 100644
--- a/cctools/misc/libtool.c
+++ b/cctools/misc/libtool.c
@@ -3172,15 +3172,17 @@ struct ofile *ofile)
 	 * Move the temporary file into its final location
 	 */
 	if (tempfile) {
-		/*
-		 * Must release mmap otherwise rename will fail on non-local fs. 
-		 * NB: the ofile_unmap can be made by caller as well
-		 */ 
-		ofile_unmap(ofile);
+	    /*
+	     * Must release mmap otherwise rename will fail on non-local fs. 
+	     * NB: the ofile_unmap can be made by caller as well
+	     */ 
+	    if (ofile) { 
+			ofile_unmap(ofile);
+		}	
 	    if (rename(tempfile, output)) {
-		system_fatal("can't move the output file to its final location: %s",
-			     output);
-		return;
+			system_fatal("can't move the output file to its final location: %s",
+			              output);
+		    return;
 	    }
 	}
 }
